zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/adhocsubprocess/ActivateAdHocSubProcessActivityTest.java

```diff
@@ -0,0 +1,220 @@
+import io.camunda.zeebe.protocol.impl.record.value.adhocsubprocess.AdHocSubProcessActivityActivationRecord;
+import io.camunda.zeebe.protocol.record.Record;
+import io.camunda.zeebe.protocol.record.intent.AdHocSubProcessActivityActivationIntent;
+import io.camunda.zeebe.protocol.record.intent.ProcessInstanceIntent;
+import io.camunda.zeebe.protocol.record.value.BpmnElementType;
+import io.camunda.zeebe.test.util.record.RecordingExporter;
+import io.camunda.zeebe.test.util.record.RecordingExporterTestWatcher;
+import java.util.List;
+import java.util.function.Consumer;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class ActivateAdHocSubProcessActivityTest {
+  @ClassRule public static final EngineRule ENGINE = EngineRule.singlePartition();
+
+  private static final String PROCESS_ID = "process";
+  private static final String AD_HOC_SUB_PROCESS_ELEMENT_ID = "ad-hoc";
+
+  @Rule public final RecordingExporterTestWatcher watcher = new RecordingExporterTestWatcher();
+
+  private long processInstanceKey;
+  private long adHocSubProcessInstanceKey;
+
+  @Before
+  public void setUp() {
+    final BpmnModelInstance processInstance =
+        process(
+            adHocSubProcess -> {
+              adHocSubProcess.task("A");
+              adHocSubProcess.task("B");
+              adHocSubProcess.task("C");
+            });
+    ENGINE.deployment().withXmlResource(processInstance).deploy();
+    processInstanceKey = ENGINE.processInstance().ofBpmnProcessId(PROCESS_ID).create();
+
+    adHocSubProcessInstanceKey =
+        RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_ACTIVATED)
+            .withProcessInstanceKey(processInstanceKey)
+            .withElementType(BpmnElementType.AD_HOC_SUB_PROCESS)
+            .map(Record::getKey)
+            .limit(1)
+            .toList()
+            .getFirst();
+  }
+
+  private BpmnModelInstance process(final Consumer<AdHocSubProcessBuilder> modifier) {
+    return Bpmn.createExecutableProcess(PROCESS_ID)
+        .startEvent()
+        .adHocSubProcess(AD_HOC_SUB_PROCESS_ELEMENT_ID, modifier)
+        .endEvent()
+        .done();
+  }
+
+  @Test
+  public void
+      givenRunningAdhocSubProcessInstanceWhenActivatingExistingFlowNodeThenTheFlowNodeIsActivated() {
+    final var adHocSubProcessActivityActivationRecord =
+        new AdHocSubProcessActivityActivationRecord()
+            .setAdHocSubProcessInstanceKey(String.valueOf(adHocSubProcessInstanceKey));
+    adHocSubProcessActivityActivationRecord.flowNodes().add().setFlowNodeId("A");
+
+    ENGINE.writeRecords(
+        RecordToWrite.command()
+            .adHocSubProcessActivityActivation(adHocSubProcessActivityActivationRecord)
+            .key(adHocSubProcessInstanceKey));
+
+    assertThat(
+            RecordingExporter.processInstanceRecords().withProcessInstanceKey(processInstanceKey))
+        // todo add short-circuit `limit` call?
```

Comment:
🔧 I think we should be able to always `.limitToProcessInstanceCompleted()` for the positive tests to avoid running into the 5 second stream timeout and having the test block for so long.